types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,207 @@
+package mempool
+
+import (
+	crand "crypto/rand" // #nosec // crypto/rand is used for seed generation
+	"encoding/binary"
+	"fmt"
+	"math/rand" // #nosec // math/rand is used for random selection and seeded from crypto/rand
+
+	huandu "github.com/huandu/skiplist"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+var (
+	_ Mempool  = (*senderNonceMempool)(nil)
+	_ Iterator = (*senderNonceMepoolIterator)(nil)
+)
+
+// senderNonceMempool is a mempool that prioritizes transactions within a sender by nonce, the lowest first,
+// but selects a random sender on each iteration.  The mempool is iterated by:
+//
+// 1) Maintaining a separate list of nonce ordered txs per sender
+// 2) For each select iteration, randomly choose a sender and pick the next nonce ordered tx from their list
+// 3) Repeat 1,2 until the mempool is exhausted
+//
+// Note that PrepareProposal could choose to stop iteration before reaching the end if maxBytes is reached.
+type senderNonceMempool struct {
+	senders map[string]*huandu.SkipList
```

Comment:
is the mempool size here unbounded?

Reply:
yes it is unbounded but this make sense to e gonna start working on a solution

Reply:
We can merge this and open another pr for that.

Reply:
Ok